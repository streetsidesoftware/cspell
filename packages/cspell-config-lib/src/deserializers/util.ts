export function detectIndent(content: string): string {
    const m = content.match(/^[ \t]+/m);
    return (m && m[0]) || '  ';
}

export function detectIndentAsNum(content: string): number {
    const indent = detectIndent(content).replace(/\t/g, '    ').replace(/[^ ]/g, '');
    return indent.length;
}
