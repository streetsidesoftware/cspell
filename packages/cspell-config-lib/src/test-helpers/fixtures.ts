import * as path from 'path';

const fixturesDir = path.join(__dirname, '../../fixtures');

export function fixtures(pathOfFixture = ''): string {
    return path.resolve(fixturesDir, pathOfFixture);
}
