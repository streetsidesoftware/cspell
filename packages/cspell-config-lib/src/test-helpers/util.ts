import { promises as fs } from 'fs';
import * as path from 'path';

export function json(obj: unknown, indent: string | number = 2): string {
    return JSON.stringify(obj, null, indent) + '\n';
}

export function tempPath(file: string): string {
    const testState = expect.getState();
    return path.join(__dirname, '../../temp', testState.currentTestName || 'test', file);
}

export async function createPathForFile(file: string): Promise<void> {
    await fs.mkdir(path.dirname(file), { recursive: true });
}

export async function copyFile(fromFile: string, toFile: string): Promise<void> {
    await createPathForFile(toFile);
    return fs.copyFile(fromFile, toFile);
}
