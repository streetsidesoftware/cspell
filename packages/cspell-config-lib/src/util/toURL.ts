export function toURL(url: string | URL): URL {
    return typeof url === 'string' ? new URL(url) : url;
}
