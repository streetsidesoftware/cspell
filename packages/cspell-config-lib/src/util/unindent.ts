/**
 * Inject values into a template string.
 */
export function inject(template: TemplateStringsArray, ...values: unknown[]): string {
    const strings = template;
    const adjValues = [];
    for (let i = 0; i < values.length; ++i) {
        const prevLines = strings[i].split('\n');
        const currLine = prevLines[prevLines.length - 1];
        const padLen = padLength(currLine);
        const padding = ' '.repeat(padLen);
        const value = `${values[i]}`;
        let pad = '';
        const valueLines = [];
        for (const line of value.split('\n')) {
            valueLines.push(pad + line);
            pad = padding;
        }
        adjValues.push(valueLines.join('\n'));
    }

    return unindent(template, ...adjValues);
}

/**
 * Calculate the padding at the start of the string.
 */
function padLength(s: string): number {
    return s.length - s.trimStart().length;
}

export function unindent(str: string): string;
export function unindent(str: TemplateStringsArray, ...values: unknown[]): string;
export function unindent(str: TemplateStringsArray | string, ...values: unknown[]): string {
    const s = typeof str === 'string' ? str : String.raw({ raw: str }, ...values);
    return _unindent(s);
}

/**
 * Remove the left padding from a multi-line string.
 */
function _unindent(str: string): string {
    const lines = str.split('\n');
    let curPad = str.length;
    for (const line of lines) {
        if (!line.trim()) continue;
        curPad = Math.min(curPad, padLength(line));
    }

    return lines.map((line) => line.slice(curPad)).join('\n');
}
