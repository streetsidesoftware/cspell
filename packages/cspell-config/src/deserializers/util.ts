export function detectIndent(content: string): string {
    const m = content.match(/^[ \t]+/m);
    return (m && m[0]) || '  ';
}
