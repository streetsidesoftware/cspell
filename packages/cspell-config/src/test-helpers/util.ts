import * as path from 'path';
import { promises as fs } from 'fs';

export function json(obj: unknown, indent: string | number = 2): string {
    return JSON.stringify(obj, null, indent) + '\n';
}

export function tempPath(file: string): string {
    const testState = expect.getState();
    return path.join(__dirname, '../../temp', testState.currentTestName, file);
}

export async function createPathForFile(file: string): Promise<void> {
    await fs.mkdir(path.dirname(file), { recursive: true });
}

export async function copyFile(fromFile: string, toFile: string): Promise<void> {
    await createPathForFile(toFile);
    return fs.copyFile(fromFile, toFile);
}
