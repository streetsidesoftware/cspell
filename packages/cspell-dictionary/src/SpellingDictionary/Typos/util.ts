import { opConcatMap, opFilter, pipe } from '@cspell/cspell-pipe/sync';
import type { TypoEntry, TyposDef, TyposDefKey, TyposDefValue, TypoValueWithSuggestions } from './typos';

/**
 * Append an entry to a TyposDef.
 * @param def - modified in place
 * @param entry- entry to add.
 * @returns def
 */
export function appendToDef(def: TyposDef, entry: TypoEntry | undefined): TyposDef {
    if (!entry) return def;
    if (typeof entry === 'string') {
        def[entry] = false;
        return def;
    }
    if (Array.isArray(entry)) {
        const [key, ...sugs] = entry.map((s) => s.trim());
        if (!key) return def;
        const s = sugs.map((s) => s.trim()).filter((s) => !!s);
        def[key] = !s.length ? false : s.length === 1 ? s[0] : s;
        return def;
    }

    Object.assign(def, entry);
    return def;
}

export function createTyposDef(entries?: Iterable<[TyposDefKey, TyposDefValue]>): TyposDef {
    const def: TyposDef = Object.create(null);

    if (!entries) return def;

    for (const [key, value] of entries) {
        def[key] = isDefined(value) ? value : false;
    }

    return def;
}

/**
 * Extract all suggestions.
 * @param typosDef - the def
 * @returns the set of suggestions.
 */
export function extractAllSuggestions(typosDef: TyposDef): Set<string> {
    const allSugs = pipe(
        Object.values(typosDef),
        opFilter(hasSuggestions),
        opConcatMap((v) => (Array.isArray(v) ? v : [v]))
    );
    return new Set(allSugs);
}

/**
 * Extract all words that have been explicitly ignore because they contains the `ignorePrefix`.
 * @param typosDef - the def
 * @param ignorePrefix - prefix
 * @returns set of ignored words with the prefix removed.
 */
export function extractIgnoreValues(typosDef: TyposDef, ignorePrefix: string): Set<string> {
    const pfxLen = ignorePrefix.length;
    return new Set(
        Object.keys(typosDef)
            .filter((k) => k.startsWith(ignorePrefix))
            .map((k) => k.slice(pfxLen))
    );
}

function isDefined<T>(v: T | undefined | null): v is T {
    return v !== undefined && v !== null;
}

function isString(v: unknown): v is string {
    return typeof v === 'string';
}

function isArray<T>(v: T[] | unknown): v is T[] {
    return Array.isArray(v);
}

function hasSuggestions(v: TyposDefValue): v is TypoValueWithSuggestions {
    return isString(v) || isArray(v);
}
