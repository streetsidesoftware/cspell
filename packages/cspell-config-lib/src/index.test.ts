import assert from 'node:assert';
import { promises as fs } from 'node:fs';
import { resolve } from 'node:path';
import { pathToFileURL } from 'node:url';

import { describe, expect, test } from 'vitest';

import * as index from './index.js';
import { createReaderWriter, CSpellConfigFile } from './index.js';
import { fixtures } from './test-helpers/fixtures.js';
import { copyFile, tempPath } from './test-helpers/util.js';

describe('index', () => {
    test('index', () => {
        expect(index).toBeDefined();
    });
    test.each`
        value                        | expected
        ${typeof createReaderWriter} | ${'function'}
    `('exports', ({ value, expected }) => {
        expect(value).toEqual(expected);
    });
});

describe('cspell-config', () => {
    test.each`
        fixture                                 | addWords
        ${'package/with-value/package.json'}    | ${['apple']}
        ${'package/without-value/package.json'} | ${['apple']}
        ${'cspell.jsonc'}                       | ${['apple', 'cache']}
        ${'cspell.yaml'}                        | ${['apple', 'cache']}
    `('edit config', async ({ fixture, addWords }) => {
        const fixtureFile = fixtures(fixture);
        const tempFile = tempPath(fixture);
        await copyFile(fixtureFile, tempFile);
        const rw = createReaderWriter();
        const uri = pathToFileURL(resolve(tempFile));
        const cfg = await rw.readConfig(uri);
        assert(cfg instanceof CSpellConfigFile);
        cfg.addWords(addWords);
        await rw.writeConfig(cfg);
        expect(await fs.readFile(tempFile, 'utf-8')).toMatchSnapshot();
    });
});
