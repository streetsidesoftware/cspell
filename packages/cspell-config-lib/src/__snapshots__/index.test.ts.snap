// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cspell-config > edit config 1`] = `
"{
  "name": "test-package",
  "cspell": {
    "words": [
      "apple",
      "banana"
    ]
  }
}
"
`;

exports[`cspell-config > edit config 2`] = `
"{
  "name": "test-package",
  "cspell": {
    "words": [
      "apple"
    ]
  }
}
"
`;

exports[`cspell-config > edit config 3`] = `
"{
    // file version
    "version": "0.2",
    "words": [
        "angle", // with comment
        "apple",
        // Before "before"
        "before",
        "cache",
        "zebra" // zebra comment
    ]
}
"
`;

exports[`cspell-config > edit config 4`] = `
"---
# file version
version: '0.2'
words:
  # First Group
  - blue
  - green
  - orange
  - red
  - yellow

  - angle # angle comment
  - apple
  - banana
  # Before "before"
  - before
  - cache # cache comment
  - claw
  - dog
  - elephant
  - zebra # zebra comment

# Last Comment
"
`;
